pkg/contractreader/observed.go

```diff
@@ -0,0 +1,159 @@
+package contractreader
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"golang.org/x/exp/maps"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/types"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query/primitives"
+)
+
+var (
+	buckets = []float64{
+		float64(10 * time.Millisecond),
+		float64(20 * time.Millisecond),
+		float64(50 * time.Millisecond),
+		float64(70 * time.Millisecond),
+		float64(100 * time.Millisecond),
+		float64(200 * time.Millisecond),
+		float64(500 * time.Millisecond),
+		float64(700 * time.Millisecond),
+		float64(time.Second),
+		float64(2 * time.Second),
+		float64(5 * time.Second),
+	}
+
+	CrDirectRequestsDurations = promauto.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "contract_reader_direct_request_duration",
+			Help:    "The amount of time elapsed during the ChainReader's GetLatestValue execution",
+			Buckets: buckets,
+		},
+		[]string{"chainID", "contract", "method"},
+	)
+	CrBatchRequestsDurations = promauto.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "contract_reader_batch_request_duration",
+			Help:    "The amount of time elapsed during the ChainReader's BatchGetLatestValues execution",
+			Buckets: buckets,
+		},
+		[]string{"chainID"},
+	)
+	CrBatchSizes = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "contract_reader_batch_request_size",
+			Help: "The size of the batch request",
+		},
+		[]string{"chainID"},
+	)
```

Comment:
Shouldn't size be a histogram as well?

Reply:
Histogram works well for bounded values (when the lower and upper limit is known). I think for case like this counter is more accurate and more lightweight, you can see:
* how often we batch and what is the average batch size
* what is the rate of batching